app/src/main/java/com/example/cleanarchitecture/viewmodel/CharacterViewModel.kt

```diff
@@ -12,29 +12,32 @@ import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 
-
 class CharacterViewModel(private val characterUseCase: CharacterUseCase) : ViewModel() {
 
     private var mutableCharactersState: MutableLiveData<Event<Data<CharacterDataWrapperEntity>>> = MutableLiveData()
     val liveDataState: LiveData<Event<Data<CharacterDataWrapperEntity>>>
         get() = mutableCharactersState
 
-    fun onFetchCharactersClicked() {
+    fun fetchCharacters() {
         viewModelScope.launch {
             mutableCharactersState.value = Event(Data(responseType = Status.LOADING))
             when (val result = withContext(Dispatchers.IO) { characterUseCase() }) {
                 is Result.Success -> {
-                    mutableCharactersState.value = Event(Data(responseType = Status.SUCCESSFUL, data = result.data))
+                    if (result.data.data.results.isNotEmpty()) {
+                        mutableCharactersState.value = Event(Data(responseType = Status.SUCCESSFUL, data = result.data))
+                    } else {
+                        mutableCharactersState.value = Event(Data(responseType = Status.EMPTY_LIST, data = result.data))
+                    }
                 }
                 is Result.Failure -> {
-                    mutableCharactersState.value = Event(Data(responseType = Status.ERROR, error = result.exception))
+                    mutableCharactersState.value = Event(Data(responseType = Status.ERROR, error = result.exception.message))
                 }
             }
         }
 
     }
 }
 
-data class Data<RequestData>(var responseType: Status, var data: RequestData? = null, var error: Exception? = null)
+data class Data<RequestData>(var responseType: Status, var data: RequestData? = null, var error: String? = null)
```

Comment:
is there any way to avoid these nullables?

Reply:
also, whats requestData? it shouldnt be `ResponseData `... or in fact, something like `CharacterResponseDate`

Reply:
to avoid nullables as nico say, i think you can conbine `data class Data<RequestData>` and `enum class Status { SUCCESSFUL, ERROR, LOADING, EMPTY_LIST }` into a new sealed class

Reply:
to avoid nullables as nico say, i think you can combine data class Data<RequestData> and enum class Status { SUCCESSFUL, ERROR, LOADING, EMPTY_LIST } into a new sealed class

Reply:
done